### PR TITLE
Rework user sessions system

### PR DESCRIPTION
**Pull request by partydragen**
Tasks:
- [x] Switch session to use hash
- [x] Logout all other hashes when user changing password and enables tfa
- [x] Update database
- [ ] Add new UserCP page to view sessions

### CODE REVIEW

**Comment by Aberdeener** [comment]
Two comments:
- "Add new UserCP page to view sessions"; this means we will require templates to update, meaning it will have to be in a major 2.x.0 update
- Is it possible to store the login method in the sessions table? Would be neat if we could display how they logged in on each of their active sessions!

**Comment by partydragen** [comment]
> Two comments:
> 
> * "Add new UserCP page to view sessions"; this means we will require templates to update, meaning it will have to be in a major 2.x.0 update
> * Is it possible to store the login method in the sessions table? Would be neat if we could display how they logged in on each of their active sessions!

1. Not really a breaking change but people need to update their templates yea, at the same time this is a security feature with make it important, Lets see what Sam says

2. Should be possible yea by using the existing method param with show username/email/oauth atleast 
3.

File: core/classes/Core/User.php [hunk 1]
```diff
@@ -307,29 +311,25 @@ public function login(?string $username = null, ?string $password = null, bool $
 
     private function _commonLogin(?string $username, ?string $password, bool $remember, string $method, bool $is_admin): bool {
         $sessionName = $is_admin ? $this->_admSessionName : $this->_sessionName;
-        if (!$username && !$password && $this->exists()) {
-            Session::put($sessionName, $this->data()->id);
+        if (!$username && $method == 'hash' && $this->exists()) {
+            // Logged in using hash from cookie
+            Session::put($sessionName, $password);
```
Comment (supercrafter100) [inline core/classes/Core/User.php:316]: Are we sure that this has to be the password and not the user id?

File: core/classes/Core/User.php [hunk 2]
```diff
@@ -107,7 +107,11 @@ public function find(string $value = null, string $field = 'id'): bool {
                 return true;
             }
 
-            $data = $this->_db->get('users', [$field, $value]);
+            if ($field != 'hash') {
+                $data = $this->_db->get('users', [$field, $value]);
+            } else {
+                $data = $this->_db->query('SELECT nl2_users.* FROM nl2_users LEFT JOIN nl2_users_session ON nl2_users.id=user_id WHERE hash = ? AND nl2_users_session.active = 1', [$value]);
```
Comment (Aberdeener) [inline core/classes/Core/User.php:113]: Space between `nl2_users.id = user_id` please 😄

File: core/init.php [hunk 3]
```diff
@@ -132,11 +132,11 @@
     // Do they need logging in (checked remember me)?
     if (Cookie::exists(Config::get('remember.cookie_name')) && !Session::exists(Config::get('session.session_name'))) {
         $hash = Cookie::get(Config::get('remember.cookie_name'));
-        $hashCheck = DB::getInstance()->get('users_session', ['hash', $hash]);
+        $hashCheck = DB::getInstance()->get('users_session', [['hash', $hash], ['active', 1]]);
```
Comment (Aberdeener) [inline core/init.php:135]: `['active', true]`

File: modules/Core/pages/user/settings.php [hunk 4]
```diff
@@ -410,7 +417,7 @@
 
                                     // Update email
                                     $user->update([
-                                        'email' => Output::getClean($_POST['email'])
+                                        'email' => $_POST['email']
```
Comment (Aberdeener) [inline modules/Core/pages/user/settings.php:420]: Why this change?

File: core/migrations/20220807142446_add_user_session_activity_columns.php [hunk 5]
```diff
@@ -0,0 +1,21 @@
+<?php
+declare(strict_types=1);
+
+use Phinx\Migration\AbstractMigration;
+
+final class AddUserSessionActivityColumns extends AbstractMigration
+{
+    public function change(): void
+    {
+        $table = $this->table('nl2_users_session');
```
Comment (Aberdeener) [inline core/migrations/20220807142446_add_user_session_activity_columns.php:10]: Can we add the "login method" column here?

**Review by tadhgboyle — CHANGES_REQUESTED** [review]

File: modules/Core/pages/user/settings.php (hunk 4 above)
Comment (partydragen) [reply modules/Core/pages/user/settings.php:420]: Well we no longer clean when saving to database ¯\_(ツ)_/¯

File: core/migrations/20220807142446_add_user_session_activity_columns.php (hunk 5 above)
Comment (partydragen) [reply core/migrations/20220807142446_add_user_session_activity_columns.php:10]: Done

File: core/classes/Core/User.php [hunk 6]
```diff
@@ -547,7 +560,9 @@ public function logout(): void {
      * Process logout if user is admin
      */
     public function admLogout(): void {
-        $this->_db->delete('users_admin_session', ['user_id', $this->data()->id]);
+        $this->_db->update('users_session', [['user_id', $this->data()->id], ['hash', Session::get($this->_admSessionName)]], [
```
Comment (samerton) [inline core/classes/Core/User.php:564]: This `admLogout` session should be deleted now

**Review by samerton — CHANGES_REQUESTED** [review]

**Review by samerton — APPROVED** [review]

**Comment by Aberdeener** [comment]
Is displaying the sessions in StaffCP delayed as well, or will that be added to this PR?

**Comment by partydragen** [comment]
> Is displaying the sessions in StaffCP delayed as well, or will that be added to this PR?

Delayed so pr is done

**Review by supercrafter100 — APPROVED** [review]

**Comment by supercrafter100** [comment]
Im guessing everything has been tested and confirmed working?

**Comment by partydragen** [comment]
> Im guessing everything has been tested and confirmed working?

Have tested yea, was not really that major changes and is only installer, login, re-auth and oauth that using login and its all using the user login function so it worked everywhere by my testing, the changes itself is pretty simple

**Review by tadhgboyle — APPROVED** [review]
